### PR TITLE
Added support to return consumed messages in a json format

### LINKED ISSUE #91: Add the possibility to log the consumer messages in a Json format.
It would be great if a new option for consumers is added, let's say `--output json` or `-J`, to log the messages consumed by the test client consumer in a json format including all the info already available in `ConsumerRecord`.

Examples:

[kcat](https://github.com/edenhill/kcat):
```JSON
{"topic":"my-topic-17256f","partition":0,"offset":0,"tstype":"create","ts":1716373253299,"broker":1,"key":null,"payload":"Hello-world - 0"}
{"topic":"my-topic-17256f","partition":0,"offset":1,"tstype":"create","ts":1716373253299,"broker":1,"key":null,"payload":"Hello-world - 1"}
```

[kaf](https://github.com/birdayz/kaf):
```JSON
{"key":"","offset":0,"partition":0,"payload":"Hello-world","timestamp":"2024-05-22T11:16:06.368Z"}
{"key":"","offset":1,"partition":0,"payload":"Hello-world","timestamp":"2024-05-22T11:16:06.399Z"}
```

### PR DESCRIPTION
**Pull request by franvila**
Added support to return the consumed messages in a json format to easy the way to collect the information from the log when running consumer in a kubernetes pod.

- Added `OUTPUT_FORMAT` parameter to consumer clients. Values `plain` (default) or `json`
- `KafkaConsumerRecord` created due `ConsumerRecord` is not serializable, so we can serialize the received message into a json string easily

Fixes #91.

### CODE REVIEW

**Comment by franvila** [comment]
@im-konge I have created this PR to solve an issue I created. I hope you don't mind I have taken the initiative to do it on advance as it would be very useful for me. In case I should not contribute to this project feel free to close the PR.

**Review by im-konge — COMMENTED** [review]
Thanks a lot for the PR and that you take the initiative to implement this! :)

I think that to have it as easy as possible, you can create a `record` class for the ConsumerRecord stuff (which you will need to parse from the message), so you can then use the ObjectMapper - similarly as it is done for `KafkaTopicDescription`.
That way, the manipulation will be IMHO easier and you will clean up the `KafkaConsumerClient` class.


For the additional configuration -> that suppose to be used for the configuration that should be passed to KafkaConsumer directly (so something that KafkaConsumer understands).
In my opinion, you can add another field to the configuration, which can be configured using environment variable (as other things) and which will default to `plain` output. So something like `MESSAGE_OUTPUT_FORMAT` or just `OUTPUT_FORMAT` ( or something like that).

WDYT?

**Comment by franvila** [comment]
> Thanks a lot for the PR and that you take the initiative to implement this! :)
> 
> I think that to have it as easy as possible, you can create a `record` class for the ConsumerRecord stuff (which you will need to parse from the message), so you can then use the ObjectMapper - similarly as it is done for `KafkaTopicDescription`. That way, the manipulation will be IMHO easier and you will clean up the `KafkaConsumerClient` class.

Thanks for that. I was thinking a lot about how to do it and I found recently the `KafkaTopicDescription` implementation that sounds more reasonable.

> For the additional configuration -> that suppose to be used for the configuration that should be passed to KafkaConsumer directly (so something that KafkaConsumer understands). In my opinion, you can add another field to the configuration, which can be configured using environment variable (as other things) and which will default to `plain` output. So something like `MESSAGE_OUTPUT_FORMAT` or just `OUTPUT_FORMAT` ( or something like that).
> 
> WDYT?

That's the other option I was thinking about. It could be clearer for users using `OUTPUT_FORMAT` variable.

File: clients/src/main/java/io/strimzi/kafka/KafkaConsumerClient.java [hunk 1]
```diff
@@ -103,15 +103,20 @@ private void checkAndReceiveMessages() {
...
+            } else {
+                LOGGER.info("Received message:");
+                LOGGER.info("\tpartition: {}", consumerRecord.partition());
+                LOGGER.info("\toffset: {}", consumerRecord.offset());
+                LOGGER.info("\tvalue: {}", consumerRecord.value());
+                if (consumerRecord.headers() != null) {
+                    LOGGER.info("\theaders: ");
+                    for (Header header : consumerRecord.headers()) {
+                        LOGGER.info("\t\tkey: {}, value: {}", header.key(), new String(header.value()));
+                    }
```
Comment (im-konge) [inline clients/src/main/java/io/strimzi/kafka/KafkaConsumerClient.java:119]: For the `plain` type of output message, I would add a method into the `KafkaConsumerRecord` as well and maybe add one more method for logging into the object, to keep it on one place and based on the settings just output the result.

So the code can then look like this:

```
for (ConsumerRecord<String, String> consumerRecord : records) {
    KafkaConsumerRecord record = KafkaConsumerRecord.parseKafkaConsumerRecord(consumerRecord);
    record.logMessage();
}
```

**Review by im-konge — COMMENTED** [review]
Just one thing, thanks a lot for this, it looks great :)

File: clients/src/main/java/io/strimzi/kafka/KafkaConsumerClient.java (hunk 1 above)
Comment (franvila) [reply clients/src/main/java/io/strimzi/kafka/KafkaConsumerClient.java:119]: Sounds pretty reasonable. Thanks!

File: clients/src/main/java/io/strimzi/kafka/KafkaConsumerClient.java [hunk 2]
```diff
@@ -103,17 +103,10 @@ private void checkAndReceiveMessages() {
...
-            if (record.headers() != null) {
-                LOGGER.info("\theaders: ");
-                for (Header header : record.headers()) {
-                    LOGGER.info("\t\tkey: {}, value: {}", header.key(), new String(header.value()));
-                }
-            }
+        for (ConsumerRecord<String, String> consumerRecord : records) {
+            KafkaConsumerRecord kafkaConsumerRecord = KafkaConsumerRecord.parseKafkaConsumerRecord(consumerRecord);
+            String log = kafkaConsumerRecord.logMessage(configuration.getOutputFormat());
+            LOGGER.info("Received message: {}", log);
```
Comment (franvila) [inline clients/src/main/java/io/strimzi/kafka/KafkaConsumerClient.java:109]: Leaving the LOGGER at KafkaConsumerClient level

**Review by im-konge — APPROVED** [review]

File: clients/src/main/java/io/strimzi/models/KafkaConsumerRecord.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    public String getTimestampType() {
+        return timestampType.name;
+    }
+
+    /**
+     * Gets key.
+     *
+     * @return the key
+     */
```
Comment (see-quick) [inline clients/src/main/java/io/strimzi/models/KafkaConsumerRecord.java:71]: Why did you implement the getters and setters here? Don't records in Java automatically provide getter methods?

**Review by see-quick — COMMENTED** [review]
One nit :) anyway looks good

File: clients/src/main/java/io/strimzi/models/KafkaConsumerRecord.java (hunk 3 above)
Comment (im-konge) [reply clients/src/main/java/io/strimzi/models/KafkaConsumerRecord.java:71]: That's for the object mapper I suppose. To create the JSON output.

Comment (franvila) [reply clients/src/main/java/io/strimzi/models/KafkaConsumerRecord.java]: Nice catch! I did it for the reason commented by Lukas. I thought it was necessary for the JSON output, but testing it, the getters are correctly provided by Java and it worked fine.

**Review by see-quick — APPROVED** [review]
LGTM, thanks @franvila 👍

**Pull request merged**